include/ginkgo/core/matrix/batch_diagonal.hpp

```diff
@@ -0,0 +1,631 @@
+          values_(exec, compute_batch_mem(this->get_size()))
+    {
+        size_type offset = 0;
+        for (size_type i = 0; i < num_duplications; ++i) {
+            exec->copy_from(input->get_executor().get(), input->get_size()[0],
+                            input->get_const_values(),
+                            this->get_values() + offset);
+            offset += input->get_size()[0];
+        }
+    }
+
+    /**
+     * Creates a BatchDiagonal matrix with the same configuration as the callers
+     * matrix.
+     *
+     * @returns a BatchDiagonal matrix with the same configuration as the
+     * caller.
+     */
+    virtual std::unique_ptr<BatchDiagonal> create_with_same_config() const
+    {
+        return BatchDiagonal::create(this->get_executor(), this->get_size());
+    }
+
+    void apply_impl(const BatchLinOp* b, BatchLinOp* x) const override;
+
+    void apply_impl(const BatchLinOp* alpha, const BatchLinOp* b,
+                    const BatchLinOp* beta, BatchLinOp* x) const override;
+
+    size_type linearize_index(const size_type batch, const size_type row) const
+        noexcept
+    {
+        if (this->get_size().stores_equal_sizes()) {
+            return row + batch * std::min(this->get_size().at(0)[0],
+                                          this->get_size().at(0)[1]);
+        } else {
+            return num_elems_per_batch_cumul_.get_const_data()[batch] + row;
+        }
+    }
+
+private:
+    Array<value_type> values_;
+    Array<size_type> num_elems_per_batch_cumul_;
+};
+
+
+/**
+ * Transforms the input matrix A and vector b according to
+ * S_L*A*S_R and S_L*b where '*' denotes matrix multiplication, and S_L and S_R
+ * are the left and right transormation matrices.
+ *
+ * @param exec  Exector to run the operation on.
+ * @param left  Left transformation matrix.
+ * @param right  Right transformation matrix.
+ * @param mtx  System matrix to be transformed.
+ * @param rhs  Right-hand side of the sytem to be transformed.
+ */
+template <typename ValueType>
+void two_sided_batch_system_transform(std::shared_ptr<const Executor> exec,
+                                      const BatchDiagonal<ValueType>* left,
+                                      const BatchDiagonal<ValueType>* right,
+                                      BatchLinOp* mtx,
+                                      BatchDense<ValueType>* rhs);
+
+
+}  // namespace matrix
+
+
+/**
+ * Creates and initializes a batch of diagonal matrices.
+ *
+ * This function first creates a temporary matrix, fills it with passed in
+ * values, and then copies the matrix to the requested backend.
+ *
+ * @tparam Matrix  matrix type to initialize
+ *
+ * @param vals  values used to initialize the batch vector
+ * @param exec  Executor associated to the vector
+ *
+ * @ingroup BatchLinOp
+ * @ingroup mat_formats
+ */
+template <typename ValueType>
+std::unique_ptr<matrix::BatchDiagonal<ValueType>> batch_diagonal_initialize(
```

Comment:
Do we need to name the function as such ? Can we not just put it in the `batch_diagonal` namespace and name the function as `initialize` ?

Reply:
We don't normally have type-specific namespaces in our public interface. What I wanted to do was add this as a specialization of `batch_initialize`, but it led to some annoying substitution debugging, IIRC. I think it can probably be done, but perhaps I could do it in a separate PR later.